src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -615,5 +598,51 @@ private void CheckRefReadOnlySymbols(MethodSymbol symbol)
                 _factory.CompilationState.ModuleBuilderOpt?.EnsureIsReadOnlyAttributeExists();
             }
         }
+
+#if DEBUG
+        private class LocalRewritingValidator : BoundTreeWalker
+        {
+            /// <summary>
+            /// Asserts that no unexpected nodes survived local rewriting.
+            /// </summary>
+            [Conditional("DEBUG")]
+            public static void Validate(BoundNode node)
+            {
+                new LocalRewritingValidator().Visit(node);
```

Comment:
Perhaps use singleton instance.

Reply:
Turns out we can't use a singleton instance after all :-S 
Doing so causes nasty concurrency issues with very surprising behaviors.

---
In reply to: [161901908](https://github.com/dotnet/roslyn/pull/23961#discussion_r161901908) [](ancestors = 161901908)